src/pineko/kfactor.py

```diff
@@ -0,0 +1,528 @@
+            # Removing 1.0 entries only for "ATLASZPT8TEVMDIST"
+            if target_dataset == "ATLASZPT8TEVMDIST":
+                centrals_kfactor = centrals_kfactor[np.where(centrals_kfactor != 1)]
+            # We need the pdf set to get the correct alpha values
+            alphas = lhapdf.mkAlphaS(pdf_set)
+            centrals_list.append(centrals_kfactor)
+            alphas_list.append(alphas)
+    # If there is only one Kfactor, this means that it contains all the factors for all the bins across all the grids.
+    elif len(cfac_paths) == 1:
+        if not cfac_paths[0].exists():
+            raise ValueError("KFactor does not exist.")
+        centrals_kfactor, pdf_set = read_kfactor(cfac_paths[0])
+        # Removing 1.0 entries only for "ATLASZPT8TEVMDIST"
+        if target_dataset == "ATLASZPT8TEVMDIST":
+            centrals_kfactor = centrals_kfactor[np.where(centrals_kfactor != 1)]
+        alphas = lhapdf.mkAlphaS(pdf_set)
+        # alpha is trivial: it is always the same
+        alphas_list = [alphas for _ in list_grids]
+        # Here I need to use the bin infos
+        centrals_list = [
+            centrals_kfactor[bin_limit[0] : bin_limit[1]] for bin_limit in bins_list
+        ]
+
+    else:
+        raise ValueError(
+            "Number of kfactors does not match the number of grids and is not one."
+        )
+    return (centrals_list, alphas_list)
+
+
+def compute_k_factor_grid(
+    grids_folder,
+    kfactor_folder,
+    yamldb_path,
+    compound_folder,
+    max_as,
+    target_folder=None,
+):
+    """Include the k_factor in the grid in order to have its associated order in the grid itself.
+
+    Parameters
+    ----------
+    grids_folder : pathlib.Path()
+        pineappl grids folder
+    kfactor_folder : pathlib.Path()
+        kfactors folder
+    yamldb_path : pathlib.Path()
+        path to the yaml file describing the dataset
+    compound_folder : pathlib.Path()
+        path to the compound folder
+    max_as : int
+        max as order
+    target_folder: pathlib.Path
+        path where store the new grid (optional)
+    """
+    import lhapdf  # pylint: disable=import-error
+
+    # With respect to the usual convention here max_as is max_as-1
+    max_as_test = max_as - 1
+    # Extracting info from yaml file
+    with open(yamldb_path, encoding="utf-8") as f:
+        yamldict = yaml.safe_load(f)
+    target_dataset = yamldict["target_dataset"]
+    operation = yamldict["operation"]
+    compound_path = pathlib.Path(compound_folder) / f"FK_{target_dataset}-COMPOUND.dat"
+    is_concatenated = False
+    list_grids = []
+    # For RATIO dataset the thing is a bit more messy
+    if operation == "RATIO":
+        is_concatenated = True
+        for op_list in yamldict["operands"]:
+            list_op_list = []
+            for op in op_list:
+                list_op_list.append(op)
+            list_grids.append(list_op_list)
+    else:
+        for op in yamldict["operands"][0]:
```

Comment:
The distinction between RATIO and non-RATIO is exactly done for this reason. So, I checked the k-factors and, for what I have seen, `operands` is always a list of lists but, when the dataset is not RATIO, it is just a trivial list of lists, i.e. the external list only contains one single list with the operands (which is the reason why in the non-RATIO case I do `for op in yamldict["operands"][0]`). In the RATIO case instead I need to be careful and take both the lists that are contained in `operands`. However you are right that I would need to do the same for ASY. Thanks!

Reply:
[f79483a](https://github.com/NNPDF/pineko/pull/75/commits/f79483ac47c3061b4d3c2db7cbb405f511117639) this should have solved the problem

Reply:
I see. Okay I'll have a closer look tomorrow but I think we can make this more general. If it's not a list of list in all cases we may need to update the yamldb file or perhaps the yaml reader settings since I suspect we can make this more general.